Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm

```diff
@@ -65,6 +69,133 @@
     return nullString();
 }
 
+void WebExtensionController::getDataRecords(OptionSet<WebExtensionDataType> types, CompletionHandler<void(Vector<Ref<WebExtensionDataRecord>>)>&& completionHandler)
+{
+    if (!m_configuration->storageIsPersistent() || types.isEmpty()) {
+        completionHandler({ });
+        return;
+    }
+
+    auto recordHolder = WebExtensionDataRecordHolder::create();
+    auto aggregator = MainRunLoopCallbackAggregator::create([recordHolder, completionHandler = WTFMove(completionHandler)]() mutable {
+        Vector<Ref<WebExtensionDataRecord>> records;
+        for (auto entry : recordHolder->recordsMap)
+            records.append(entry.value);
+
+        completionHandler(records);
+    });
+
+    auto uniqueIdentifiers = FileSystem::listDirectory(m_configuration->storageDirectory());
+    for (auto& uniqueIdentifier : uniqueIdentifiers) {
+        String displayName;
+        URL lastBaseURL;
+        if (!WebExtensionContext::readDisplayNameAndLastBaseURLFromState(stateFilePath(uniqueIdentifier), displayName, lastBaseURL))
+            continue;
+
+        for (auto type : types) {
+            auto *storage = sqliteStore(storageDirectory(uniqueIdentifier), type, this->extensionContext(lastBaseURL));
+            if (!storage)
```

Comment:
actually don't want to continue here since DOM and indexedDB storage can still be removed. will fix.